gk/main.c

```diff
@@ -648,6 +576,143 @@ alloc_gk_conf(void)
+
+	hint.ai_family = PF_UNSPEC;
+	hint.ai_flags = AI_NUMERICHOST;
+
+	ret = getaddrinfo(ip_addr, NULL, &hint, &res);
+    	if (ret) {
+        	RTE_LOG(ERR, GATEKEEPER,
+			"gk: invalid ip address %s; %s\n",
+			ip_addr, gai_strerror(ret));
+        	return 1;
+    	}
+
+    	if(res->ai_family != AF_INET && res->ai_family != AF_INET6)
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: %s is an is unknown address format %d\n",
+			ip_addr, res->ai_family);
+
+	ret = res->ai_family;
+   	freeaddrinfo(res);
+
+	return ret;
+}
+
+int
+lua_init_tunnel(struct gk_config *gk_conf, const char *gw,
+	const char **gk_addrs, uint8_t num_gk_addrs,
+	const char **gt_addrs, uint8_t num_gt_addrs)
+{
+	int i;
+	int ret;
+	struct ether_addr gateway;
+
+	memset(&gk_conf->gk_inaddr, 0, sizeof(struct in_addr));
+	memset(&gk_conf->gk_inaddr6, 0, sizeof(struct in6_addr));
+
+	if (sscanf(gw, "%hhx:%hhx:%hhx:%hhx:%hhx:%hhx",
+           		&gateway.addr_bytes[0],
+           		&gateway.addr_bytes[1],
+           		&gateway.addr_bytes[2],
+           		&gateway.addr_bytes[3],
+           		&gateway.addr_bytes[4],
+           		&gateway.addr_bytes[5]) != sizeof(struct ether_addr)) {
+    		RTE_LOG(ERR, GATEKEEPER,
+			"gk: cannot parse gateway MAC address %s\n",
+			gw);
+		ret = -1;
+		goto out;
+	}
+
+	for (i = 0; i < num_gk_addrs; i++) {
+		int gk_type = get_ip_type(gk_addrs[i]);
+		if (gk_type == AF_INET) {
+			if ((ret = inet_pton(AF_INET, gk_addrs[i],
+					&gk_conf->gk_inaddr)) != 1) {
+				ret = -1;
+				goto out;
+			}
+		} else if (gk_type == AF_INET6) {
+			if ((ret = inet_pton(AF_INET6, gk_addrs[i],
+					&gk_conf->gk_inaddr6)) != 1) {
+				ret = -1;
+				goto out;
+			}
+		} else {
+			ret = -1;
+			goto out;
+		}
+	}
+
+	gk_conf->num_tunnels = num_gt_addrs;
+	gk_conf->tunnels = rte_calloc("ipip_tunnel",
+			num_gt_addrs, sizeof(struct ipip_tunnel_info), 0);
+	if (gk_conf->tunnels == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	for (i = 0; i < num_gt_addrs; i++) {
+		struct in_addr  gt_inaddr;
+		struct in6_addr gt_inaddr6;
+		int gt_type = get_ip_type(gt_addrs[i]);
+		if (gt_type == AF_INET) {
+			if ((ret = inet_pton(AF_INET,
+					gt_addrs[i], &gt_inaddr)) != 1) {
+				ret = -1;
+				goto tunnels;
+			}
+
+			gk_conf->tunnels[i].flow.proto = ETHER_TYPE_IPv4;
+			gk_conf->tunnels[i].flow.f.v4.src =
+				gk_conf->gk_inaddr.s_addr;
+			gk_conf->tunnels[i].flow.f.v4.dst =
+				gt_inaddr.s_addr;
+		} else if (gt_type == AF_INET6) {
+			if ((ret = inet_pton(AF_INET6,
+					gt_addrs[i], &gt_inaddr6)) != 1) {
+				ret = -1;
+				goto tunnels;
+			}
+
+			gk_conf->tunnels[i].flow.proto = ETHER_TYPE_IPv6;
+			rte_memcpy(gk_conf->tunnels[i].flow.f.v6.src,
+					gk_conf->gk_inaddr6.s6_addr,
+					sizeof(gk_conf->tunnels[i].flow.f.v6.src));
+			rte_memcpy(gk_conf->tunnels[i].flow.f.v6.dst,
+					gt_inaddr6.s6_addr,
+					sizeof(gk_conf->tunnels[i].flow.f.v6.dst));
+		} else {
+			ret = -1;
+			goto tunnels;
+		}
+
+		ether_addr_copy(&gateway, &gk_conf->tunnels[i].gateway_mac);
```

Comment:
Can't the gateway address be found as the Ethernet address of the back interface, using `struct net_config`?